### PR TITLE
Cleanup OAuth, use clean interceptor pattern

### PR DESCRIPTION
**Pull request by daveta**
Cleans up alot of stuff, abandon C# style token resolve, and make fully async.

### CODE REVIEW

File: libraries/bot-connector/src/main/java/com/microsoft/bot/connector/authentication/OAuthClient.java [hunk 1]
```diff
@@ -84,54 +89,85 @@ public URI MakeUri(String uri, HashMap<String, String> queryStrings) throws URIS
...
-        }
-        else {
-            return completedFuture(null);
-        }
+        return CompletableFuture.supplyAsync(() -> {
+            // Construct URL
+            HashMap <String, String> qstrings = new HashMap<>();
+            qstrings.put("userId", userId);
+            qstrings.put("connectionName", connectionName);
+            if (!StringUtils.isBlank(magicCode))
```
Comment (johnataylor) [inline libraries/bot-connector/src/main/java/com/microsoft/bot/connector/authentication/OAuthClient.java]: single line if

File: libraries/bot-connector/src/main/java/com/microsoft/bot/connector/authentication/OAuthClient.java [hunk 2]
```diff
@@ -84,54 +89,85 @@ public URI MakeUri(String uri, HashMap<String, String> queryStrings) throws URIS
...
+                    return this.mapper.readValue(response.body().string(), TokenResponse.class);
+                }
+                else if (statusCode == HTTP_NOT_FOUND) {
+                    return null;
+                }
+                else {
+                    return null;
+                }
+            } catch (IOException e) {
+                e.printStackTrace();
```
Comment (johnataylor) [inline libraries/bot-connector/src/main/java/com/microsoft/bot/connector/authentication/OAuthClient.java]: is eating the exception intended?

File: libraries/bot-connector/src/main/java/com/microsoft/bot/connector/authentication/OAuthClient.java [hunk 3]
```diff
@@ -141,85 +177,124 @@ else if (statusCode == HTTP_NOT_FOUND) {
...
+                    .build();
+
+            Response response = null;
+            try {
+                response = client.newCall(request).execute();
+                int statusCode = response.code();
+                if (statusCode == HTTP_OK)
+                    return response.body().string();
+            } catch (IOException e) {
+                e.printStackTrace();
```
Comment (johnataylor) [inline libraries/bot-connector/src/main/java/com/microsoft/bot/connector/authentication/OAuthClient.java]: eat exception intended?

File: libraries/bot-connector/src/main/java/com/microsoft/bot/connector/authentication/OAuthClient.java [hunk 4]
```diff
@@ -232,14 +307,38 @@ public CompletableFuture SendEmulateOAuthCardsAsync(Boolean emulateOAuthCards) t
...
+                    .build();
+
+            Response response = null;
+            try {
+                response = client.newCall(request).execute();
+                int statusCode = response.code();
+                if (statusCode == HTTP_OK)
+                    return;
+            } catch (IOException e) {
+                e.printStackTrace();
```
Comment (johnataylor) [inline libraries/bot-connector/src/main/java/com/microsoft/bot/connector/authentication/OAuthClient.java]: eat exception?

File: libraries/bot-connector/src/main/java/com/microsoft/bot/connector/authentication/OAuthClient.java (hunk 1 above)
Comment (daveta) [reply libraries/bot-connector/src/main/java/com/microsoft/bot/connector/authentication/OAuthClient.java]: Fixed

File: libraries/bot-connector/src/main/java/com/microsoft/bot/connector/authentication/OAuthClient.java (hunk 2 above)
Comment (daveta) [reply libraries/bot-connector/src/main/java/com/microsoft/bot/connector/authentication/OAuthClient.java]: Yes, failure condition to caller

File: libraries/bot-connector/src/main/java/com/microsoft/bot/connector/authentication/OAuthClient.java (hunk 3 above)
Comment (daveta) [reply libraries/bot-connector/src/main/java/com/microsoft/bot/connector/authentication/OAuthClient.java]: Yeah, hard to get context - these are retrieving strings and tokens and stuff.  If it fails, just returns null or false in one case.

**Pull request merged**